Implement Trace(AB) using vdot for performance efficiency

thisisRMak opened the pull request:
## Description
Implement Trace(AB) using vdot for performance efficiency. Move trace to trace_internal and use trace as a wrapper function that determines when to use vdot for a faster computation `vdot(A,B)` versus fall back to traditional trace `trace_internal`
Reference: https://github.com/cvxpy/cvxpy/discussions/2359
In experimental testing we see 
- 7x improvement on canonicalization of the following problem reducing execution time from 7.46 seconds to 1.05 seconds
- We also see 10x improvement in ``semidefinite_programming.SemidefiniteProgramming.time_compile_problem``
```py
import numpy as np
import cvxpy as cp
import time

# --- Step 1: Generate synthetic data ---
np.random.seed(0)
n = 500     # dimension (number of variables)
N = 100    # number of samples

# Create a random positive definite covariance matrix
A = np.random.randn(n, n)
true_cov = A @ A.T + n * np.eye(n)  # SPD matrix
X = np.random.multivariate_normal(mean=np.zeros(n), cov=true_cov, size=N)

# Sample covariance
S = np.cov(X, rowvar=False)

# --- Step 2: Define Graphical Lasso problem ---
lambda_reg = 0.1
K = cp.Variable((n, n), symmetric=True)

# Objective: -log det(K) + trace(SK) + λ * ||K||_1
log_det_term = -cp.log_det(K)
trace_term   = cp.trace(S @ K)
l1_penalty   = lambda_reg * cp.norm1(K)

objective = cp.Minimize(log_det_term + trace_term + l1_penalty)

constraints = [K >> 0]  # enforce K ≽ 0

prob = cp.Problem(objective, constraints)
start_time = time.perf_counter()
data, _, _ = prob.get_problem_data(solver=cp.CLARABEL)
end_time = time.perf_counter()
elapsed_time = end_time - start_time
print(f"Function execution time: {elapsed_time:.6f} seconds")
```

## Type of change
- [x] New feature (backwards compatible)
- [ ] New feature (breaking API changes)
- [ ] Bug fix
- [ ] Other (Documentation, CI, ...)

## [Contribution checklist](https://www.cvxpy.org/contributing/index.html#contribution-checklist)
- [ ] Add our license to new files.
- [x] Check that your code adheres to our coding style.
- [x] Write unittests.
- [x] Run the unittests and check that they’re passing.
- [ ] Run the benchmarks to make sure your change doesn’t introduce a regression.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/cvxpy/cvxpy?pullRequest=2920) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/cvxpy/cvxpy?pullRequest=2920) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/cvxpy/cvxpy?pullRequest=2920) it.</sub>

PTNobel commented:
Thanks for the contribution! We synced synchronously; I'll review once you have the new design.

PTNobel commented:
I think this usage also needs to be fixed.

Can we also test some complex matrices? I'm a bit worried that the vdot conj is incorrect

Transurgeon commented:
> I think this usage also needs to be fixed:
> 
> https://github.com/cvxpy/cvxpy/blob/9e51b84e3130d653a16efa5c8b80cd82a035eb01/cvxpy/reductions/dgp2dcp/canonicalizers/__init__.py#L7
> 
> Can we also test some complex matrices? I'm a bit worried that the vdot conj is incorrect

It is testing complex matrices already. The tests pass for TestComplex and also conic solver tests have a complex test case.
We noticed this because there were 4 failing tests without the change to the complex2real reduction.

jonathanberthias commented:
Naming wise, what would you think of renaming `trace_internal` to `Trace` (with a capital T)? It's a pattern that is already used a little in the code base. Off the top of my head, there's `sum`/`Sum` and `hstack`/`Hstack`, probably a few others.

thisisRMak commented:
Thanks @PTNobel @Transurgeon @jonathanberthias . 
- [x] trace_internal is now renamed to Trace
- [x] added trace_test_dgp2dcp in test_atoms.py
  - Does it match what you had in mind @PTNobel ?
  - Do we want `trace_test_dgp2dcp` placed in `test_dgp.py` instead of `test_atoms.py`?
- [ ] Do we want a similar test added for `test_trace_complex2real`? 
  - I've verified this test case fails without our complex2real fix, but passes with it. I've pushed this change also so we can see if passing workflows, in case the group wants to include this. Plmk if we prefer it to be placed in `test_complex.py` instead of `test_atoms.py` as I've implemented it now?
  - Like @Transurgeon mentioned earlier, trace under complex2real functionality was already getting exercised under pytest, since these 4 tests failed earlier without the change to complex2real's canonicalizer. Now `test_trace_complex2real` makes it 5 test failures, but maybe its helpful if it explicitly pins the blame on trace when these tests fail together.

```
FAILED cvxpy/tests/test_complex.py::TestComplex::test_log_det - AssertionError
FAILED cvxpy/tests/test_complex.py::TestComplex::test_special_idx - AssertionError
FAILED cvxpy/tests/test_conic_solvers.py::TestSCS::test_complex_matrices - AssertionError
FAILED cvxpy/tests/test_von_neumann_entr.py::Test_von_neumann_entr::test_1_complex - AssertionError
```

thisisRMak commented:
There was a bug in test_trace_complex2real. Minimize needs a "Real" target. Added a norm like some other tests in test_complex.py seem to be doing. Wasn't a scientific decision, just poking around to see something that works. 

Suggestions welcome to either drop test_trace_complex2real completely or to do a different test than what I've hacked together here.
